include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -50,13 +52,35 @@ class device_policy
     device_policy(const device_policy<OtherName>& other) : q(other.queue())
     {
     }
+
+    device_policy(const device_policy& other) : q(other.queue()) {}
+    device_policy(device_policy&& other) : q(other.queue()) {}
+    device_policy&
+    operator=(const device_policy& other)
+    {
+        q = other.queue();
+        return *this;
+    }
+    device_policy&
+    operator=(device_policy&& other)
+    {
+        q.swap(other.q);
+        return *this;
+    }
+
     explicit device_policy(sycl::queue q_) : q(q_) {}
-    explicit device_policy(sycl::device d_) : q(d_) {}
-    operator sycl::queue() const { return q; }
+    explicit device_policy(sycl::device d_) { q.emplace(d_); }
+    operator sycl::queue() const { return queue(); }
     sycl::queue
     queue() const
     {
-        return q;
+        if (!q)
```

Comment:
As form it's looks like not thread-safe operation too.
The problem is in this condition, I think:
- thread (1) make call of this function and because ```if (!q)``` is true -> started to create queue under mutex.
   On last step of this process it's modify ```q``` state
- thread (2) make call of this function - it's check ```q``` state without synchronization (error 1) and may return broken state of ```q``` (error 2) while thread 1 not finished modification of ```q``` state.

Reply:
I suppose it depends on whether `std::optional<T>::emplace` can leave the optional in a state where it reports that it has a value before the value is actually available.

The alternative would be to move the lock to the outer scope, which would make the `queue()` method always lock. I would like to avoid that if possible.

Reply:
I think https://en.cppreference.com/w/cpp/utility/optional class is not thread-safe.

Reply:
Looks like Sergey is right and the double check locking pattern (DCLP) is not safe when applied to `std::optional`. Even if the implementation sets the "has a value" flag after assigning the value, it's not enough as a compiler or a processor can reorder the operations. To prevent that, a memory fence is needed, and I doubt that any (not even talking about all) implementations of `optional` does that,
So my original suggestion of DCLP does not work here, and in lieu of enough time let's just always take the lock.

Reply:
Okay, I changed it to always grab the lock now.